eyecite/count.py

```diff
@@ -0,0 +1,180 @@
+    citation: SupraCitation, opinions: List[Opinion]
+) -> Optional[Opinion]:
+    """
+    Try to match supra citations by checking whether their antecedent_guess
+    appears in either the defendant or plaintiff field of the citation of any
+    of the previously matched opinions.
+    """
+    if not citation.antecedent_guess:  # If no guess, can't do anything
+        return None
+
+    candidates: List[Opinion] = []
+    ag: str = strip_punct(citation.antecedent_guess)
+
+    for o in opinions:
+        if o.citation.defendant and ag in o.citation.defendant:
+            candidates.append(o)
+        elif o.citation.plaintiff and ag in o.citation.plaintiff:
+            candidates.append(o)
+
+    # Remove duplicates and only accept if one candidate remains
+    candidates = list(set(candidates))
+    return candidates[0] if len(candidates) == 1 else None
+
+
+def _match_shortcase_citation(
+    citation: ShortCaseCitation, opinions: List[Opinion]
+) -> Optional[Opinion]:
+    """
+    Try to match shortcase citations by checking if their reporter and volume
+    number matches any of the previously matched opinions. If there are
+    multiple possible matches, try to refine by also checking whether their
+    antecedent_guess appears in either the defendant or plaintiff field of the
+    citation of any of the previously matched opinions.
+    """
+    candidates: List[Opinion] = []
+    for o in opinions:
+        if (
+            citation.reporter == o.citation.reporter
+            and citation.volume == o.citation.volume
+        ):
+            candidates.append(o)
+
+    # Remove duplicates and only accept if one candidate remains
+    candidates = list(set(candidates))
+    if len(candidates) == 1:
+        return candidates[0]
+
+    # Otherwise, if there is an antecedent guess, try to refine further
+    elif citation.antecedent_guess:
+        refined_candidates: List[Opinion] = []
+        ag: str = strip_punct(citation.antecedent_guess)
+
+        for c in candidates:
+            if c.citation.defendant and ag in c.citation.defendant:
+                refined_candidates.append(c)
+            elif c.citation.plaintiff and ag in c.citation.plaintiff:
+                refined_candidates.append(c)
+
+        # Remove duplicates and only accept if one candidate remains
+        refined_candidates = list(set(refined_candidates))
+        return refined_candidates[0] if len(refined_candidates) == 1 else None
+
+    # Otherwise, nothing left to try
+    else:
+        return None
+
+
+def _match_fullcase_citation(citation: FullCaseCitation) -> Opinion:
+    """
+    Use fullcase citations as direct opinion matches.
+    """
+    return Opinion(citation)
+
+
+def count_citations(
+    citations: List[CitationBase],
+    match_id_citation: Callable[
+        [IdCitation, List[Any], bool], Optional[OpinionBase]
+    ] = _match_id_citation,
+    match_supra_citation: Callable[
+        [SupraCitation, List[Any]], Optional[OpinionBase]
```

Comment:
For all of these, the call signature that I really want is `Callable[[SupraCitation, List[OpinionBase]], Optional[OpinionBase]]`, i.e., I want to specify a function that takes a supra citation (or whatever) and a list of opinions as arguments, and then optionally returns an opinion. In theory, this would then allow the user to use any custom opinion object that inherits from `OpinionBase` here.

This works fine for the _return_ type -- eyecite's own matching functions that I have written (which all return `Opinion`, a subtype of `OpinionBase`) all work without a problem.

However, this doesn't work for the _argument_ types. This is because arguments in `Callable` are apparently treated contravariantly in Python, not covariantly. See https://www.python.org/dev/peps/pep-0483/#covariance-and-contravariance for a discussion of this. Thus, the expected type inheritance is "reversed", and objects that are subtypes of `OpinionBase` are NOT accepted as arguments in a `Callable` so specified.

This seems wrong to me, but who I am to question Python's type design? In any case, my workaround is to write `List[Any]` instead of the more narrow (and more correct, imo) `List[OpinionBase]` in the `Callable` here.